contracts/axelarnet-gateway/src/state.rs

```diff
@@ -0,0 +1,235 @@
+use axelar_wasm_std::counter::Counter;
+use axelar_wasm_std::{FnExt, IntoContractError};
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Addr, StdError, Storage};
+use cw_storage_plus::{Item, Map};
+use router_api::{ChainName, CrossChainId, Message};
+
+#[cw_serde]
+pub(crate) struct Config {
+    pub chain_name: ChainName,
+    pub router: Addr,
+}
+
+#[cw_serde]
+pub(crate) enum MessageStatus {
+    Approved,
+    Executed,
+}
+
+#[cw_serde]
+pub(crate) struct MessageWithStatus {
+    pub msg: Message,
+    pub status: MessageStatus,
+}
+
+const CONFIG_NAME: &str = "config";
+const CONFIG: Item<Config> = Item::new(CONFIG_NAME);
+
+const COUNTER_NAME: &str = "counter";
+const COUNTER: Counter<u32> = Counter::new(COUNTER_NAME);
+
+const INCOMING_MESSAGES_NAME: &str = "incoming_messages";
+const INCOMING_MESSAGES: Map<CrossChainId, Message> = Map::new(INCOMING_MESSAGES_NAME);
+
+const OUTGOING_MESSAGES_NAME: &str = "outgoing_messages";
+const OUTGOING_MESSAGES: Map<CrossChainId, MessageWithStatus> = Map::new(OUTGOING_MESSAGES_NAME);
+
+#[derive(thiserror::Error, Debug, IntoContractError)]
+pub enum Error {
+    #[error(transparent)]
+    Std(#[from] StdError),
+    #[error("gateway got into an invalid state, its config is missing")]
+    MissingConfig,
+    #[error("message with ID {0} mismatches with the stored one")]
+    MessageMismatch(CrossChainId),
+    #[error("message with ID {0} not found")]
+    MessageNotFound(CrossChainId),
+    #[error("message with ID {0} not approved")]
+    MessageNotApproved(CrossChainId),
+    #[error("message with ID {0} already executed")]
+    MessageAlreadyExecuted(CrossChainId),
+    #[error("incoming message with ID {0} already exists")]
+    MessageAlreadyExists(CrossChainId),
+}
+
+pub(crate) fn save_config(storage: &mut dyn Storage, value: &Config) -> Result<(), Error> {
+    CONFIG.save(storage, value).map_err(Error::from)
+}
+
+pub(crate) fn load_config(storage: &dyn Storage) -> Result<Config, Error> {
+    CONFIG
+        .may_load(storage)
+        .map_err(Error::from)?
+        .ok_or(Error::MissingConfig)
+}
+
+pub(crate) fn save_incoming_msg(
+    storage: &mut dyn Storage,
+    key: CrossChainId,
+    msg: &Message,
+) -> Result<(), Error> {
+    match INCOMING_MESSAGES.may_load(storage, key.clone())? {
+        Some(_) => Err(Error::MessageAlreadyExists(key)),
+        None => INCOMING_MESSAGES
+            .save(storage, key, msg)
+            .map_err(Error::from),
+    }
+}
+
+pub(crate) fn may_load_incoming_msg(
+    storage: &dyn Storage,
+    id: &CrossChainId,
+) -> Result<Option<Message>, Error> {
+    INCOMING_MESSAGES
+        .may_load(storage, id.clone())
+        .map_err(Error::from)
+}
+
+#[cfg(test)]
+pub(crate) fn may_load_outgoing_msg(
+    storage: &dyn Storage,
+    cc_id: &CrossChainId,
+) -> Result<Option<MessageWithStatus>, Error> {
+    OUTGOING_MESSAGES
+        .may_load(storage, cc_id.clone())
+        .map_err(Error::from)
+}
+
+pub(crate) fn save_outgoing_msg(
```

Comment:
rename these functions? save_received and save_sent

Reply:
ah, forgot to update state.rs